internal/auth/providers/azure_graph.go

```diff
@@ -0,0 +1,170 @@
+)
+
+// AzureGroupCacheSize controls the size of the caches of AD group info
+const AzureGroupCacheSize = 1024
+
+// GraphService wraps calls to provider admin APIs
+type GraphService interface {
+	GetGroups(string) ([]string, error)
+}
+
+// AzureGraphService implements graph API calls for the Azure provider
+type AzureGraphService struct {
+	client               *http.Client
+	groupMembershipCache *lru.Cache
+	groupNameCache       *lru.Cache
+}
+
+// NewAzureGraphService creates a new graph service for getting groups
+func NewAzureGraphService(clientID string, clientSecret string, tokenURL string) *AzureGraphService {
+	clientConfig := &clientcredentials.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		TokenURL:     tokenURL,
+		Scopes: []string{
+			"https://graph.microsoft.com/.default",
+		},
+	}
+	ctx := context.Background()
+	client := clientConfig.Client(ctx)
+	memberCache, err := lru.New(AzureGroupCacheSize)
+	if err != nil {
+		panic(err) // Should only happen if AzureGroupCacheSize is a negative number
+	}
+	nameCache, err := lru.New(AzureGroupCacheSize)
+	if err != nil {
+		panic(err) // Should only happen if AzureGroupCacheSize is a negative number
+	}
+	return &AzureGraphService{
+		client:               client,
+		groupMembershipCache: memberCache,
+		groupNameCache:       nameCache,
+	}
+}
+
+// GetGroups lists groups user belongs to.
+func (gs *AzureGraphService) GetGroups(email string) ([]string, error) {
+	if gs.client == nil {
+		return []string{}, errors.New("oauth client must be configured")
+	}
+	if email == "" {
+		return []string{}, errors.New("missing email")
+	}
+
+	var wg sync.WaitGroup
+	var mtx sync.Mutex
+	var err error
+	groupNames := make([]string, 0)
+	// See: https://developer.microsoft.com/en-us/graph/docs/api-reference/beta/api/user_getmembergroups
+	requestBody := `{"securityEnabledOnly": false}`
+	requestURL := fmt.Sprintf("https://graph.microsoft.com/beta/users/%s/getMemberGroups", url.PathEscape(email))
+	for {
+		groupResponse, err := gs.client.Post(requestURL, "application/json", strings.NewReader(requestBody))
+		if err != nil {
+			return []string{}, err
+		}
+
+		groupData := struct {
+			Next  string   `json:"@odata.nextLink"`
+			Value []string `json:"value"`
+		}{}
+
+		body, err := ioutil.ReadAll(groupResponse.Body)
+		if err != nil {
+			return []string{}, err
+		}
+		if groupResponse.StatusCode >= 400 {
+			return []string{}, fmt.Errorf("api error: %s", string(body))
+		}
+
+		err = json.Unmarshal(body, &groupData)
+		if err != nil {
+			return []string{}, err
+		}
+
+		for _, groupID := range groupData.Value {
+			wg.Add(1)
+			id := groupID
+			go func(wg *sync.WaitGroup) {
+				defer wg.Done()
+
+				var name string
+				// check the cache for the group name first
+				cachedName, ok := gs.groupNameCache.Get(id)
+				if !ok {
+					// didn't have the group name, make concurrent API call to fetch it
+					name, err = gs.getGroupName(id)
+					if err == nil {
```

Comment:
Just sanity-checking: Anything we should be doing here if `err != nil`?

Reply:
TBQH, I don't remember anymore. I don't think I'd have made the mistake of just swallowing the error unless I intended to, but if so, that deserves a comment. I'll reinvestigate.

Reply:
Looks like this was intentional, but really needed comments. Added.